genie-web/src/main/java/com/netflix/genie/web/services/impl/JobResolverServiceImpl.java

```diff
@@ -672,4 +724,214 @@ private int resolveJobMemory(final JobRequest jobRequest, final Command command)
             .getRequestedJobMemory()
             .orElse(command.getMemory().orElse(this.defaultMemory));
     }
+
+    private boolean useV4ResourceSelection() {
+        // TODO: Flesh this out with dynamic behavior to aid migration between V3 and V4
+        return false;
+    }
+
+    private Command resolveCommand(final JobRequest jobRequest) throws GenieJobResolutionException {
+        final long start = System.nanoTime();
+        final Set<Tag> tags = Sets.newHashSet();
+        try {
+            final Criterion criterion = jobRequest.getCriteria().getCommandCriterion();
+            final Set<Command> commands = this.commandPersistenceService.findCommandsMatchingCriterion(criterion);
+            final Command command;
+            if (commands.isEmpty()) {
+                throw new GenieJobResolutionException("No command matching command criterion found");
+            } else if (commands.size() == 1) {
+                command = commands
+                    .stream()
+                    .findFirst()
+                    .orElseThrow(() -> new GenieJobResolutionException("No command matching criterion found."));
+                log.debug("Found single command {} matching criterion {}", command.getId(), criterion);
+            } else {
```

Comment:
I'm wondering if we should always defer to the selector even if the matching set contains a single element.
2 reasons:
 - Selector may make "smart" decision and actually not like the only match
 - Metrics and other things would be more uniform (vs. now we short-circuit in some cases).

There is a performance cost, but it's probably negligible..